lib/rubocop/cop/rspec/return_from_stub.rb

```diff
@@ -0,0 +1,84 @@
+# frozen_string_literal: true
+
+module RuboCop
+  module Cop
+    module RSpec
+      # Checks for consistent style of stub's return setting.
+      #
+      # Enforces either `and_return` or block-style return in the cases
+      # where the returned value is constant. Ignores dynamic returned values
+      # are the result would be different
+      #
+      # This cop can be configured using the `EnforcedStyle` option
+      #
+      # @example `EncorcedStyle: block`
+      #   # bad
+      #   allow(Foo).to receive(:bar).and_return("baz")
+      #   expect(Foo).to receive(:bar).and_return("baz")
+      #
+      #   # good
+      #   allow(Foo).to receive(:bar) { "baz" }
+      #   expect(Foo).to receive(:bar) { "baz" }
+      #   # also good as the returned value is dynamic
+      #   allow(Foo).to receive(:bar).and_return(bar.baz)
+      #
+      # @example `EncorcedStyle: and_return`
+      #   # bad
+      #   allow(Foo).to receive(:bar) { "baz" }
+      #   expect(Foo).to receive(:bar) { "baz" }
+      #
+      #   # good
+      #   allow(Foo).to receive(:bar).and_return("baz")
+      #   expect(Foo).to receive(:bar).and_return("baz")
+      #   # also good as the returned value is dynamic
+      #   allow(Foo).to receive(:bar) { bar.baz }
+      #
+      class ReturnFromStub < Cop
+        include ConfigurableEnforcedStyle
+
+        MSG_AND_RETURN = 'Use `and_return` for static values.'.freeze
+        MSG_BLOCK = 'Use block for static values.'.freeze
+
+        def_node_matcher :receive_with_block, <<-PATTERN
+          (block
+            (send nil :receive ...)
+            (args)
+            $(...)
+          )
+        PATTERN
+
+        def_node_matcher :and_return_value, <<-PATTERN
+            (send
+              (send nil :receive (...)) :and_return $(...)
+            )
+        PATTERN
+
+        def on_block(node)
+          return unless style == :and_return
+          receive_with_block(node) do |args|
+            add_offense(node, :expression, MSG_AND_RETURN) unless dynamic?(args)
+          end
+        end
+
+        def on_send(node)
+          return unless style == :block
+
+          and_return_value(node) do |args|
+            add_offense(node, :expression, MSG_BLOCK) unless dynamic?(args)
+          end
+        end
+
+        private
+
+        def dynamic?(node)
```

Comment:
We should probably have this method instead check to see if the value only contains literals. The `YodaCondition` cop had a lot of issues because it tried to blacklist dynamic nodes instead of whitelist static nodes: https://github.com/bbatsov/rubocop/issues?q=is%3Aissue+YodaCondition+label%3Abug